parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java

```diff
@@ -66,19 +63,21 @@
   @SuppressWarnings("unchecked")
   ReadConf(InputFile file, ParquetReadOptions options, Schema expectedSchema, Expression filter,
            Function<MessageType, ParquetValueReader<?>> readerFunc, Function<MessageType,
-           VectorizedReader<?>> batchedReaderFunc, boolean reuseContainers,
+           VectorizedReader<?>> batchedReaderFunc, NameMapping nameMapping, boolean reuseContainers,
            boolean caseSensitive, Integer bSize) {
     this.file = file;
     this.options = options;
     this.reader = newReader(file, options);
     MessageType fileSchema = reader.getFileMetaData().getSchema();
-
     boolean hasIds = ParquetSchemaUtil.hasIds(fileSchema);
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
+    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema, nameMapping);
+
+    if (nameMapping == null && !hasIds) {
```

Comment:
This logic assumes what `addFallbackIds` is going to do. I think I'd rather structure this a bit differently:

```
MessageType typeWithIds;
if (ParquetSchemaUtil.hasIds(fileSchema)) {
  typeWithIds = fileSchema;
  this.projection = ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema);
} else if (nameMapping != null) {
  typeWithIds = ParquetSchemaUtil.applyNameMapping(fileSchema, nameMapping);
  this.projection = ParquetSchemaUtil.pruneColumns(typeWithIds, expectedSchema);
} else {
  typeWithIds = ParquetSchemaUtil.addFallbackIds(fileSchema);
  this.projection = ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
}
```

Reply:
Done.